src/app/qgsmaptoolscalefeature.cpp

```diff
@@ -0,0 +1,479 @@
+        return;
+      }
+
+      mExtent = cf.geometry().boundingBox();
+      mFeatureCenter = mExtent.center();
+
+      mScaledFeatures.clear();
+      mScaledFeatures << cf.id(); //todo: take the closest feature, not the first one...
+
+      mRubberBand = createRubberBand( mLayer->geometryType() );
+      mRubberBand->setToGeometry( cf.geometry(), mLayer );
+    }
+    else
+    {
+      mScaledFeatures = mLayer->selectedFeatureIds();
+
+      mRubberBand = createRubberBand( mLayer->geometryType() );
+
+      QgsFeature feat;
+      QgsFeatureIterator it = mLayer->getSelectedFeatures();
+      while ( it.nextFeature( feat ) )
+      {
+        mRubberBand->addGeometry( feat.geometry(), mLayer );
+      }
+    }
+    QgsPointXY mapAnchor = toMapCoordinates( mLayer,mFeatureCenter );
+    QPoint rubberAnchor = toCanvasCoordinates( mapAnchor );
+    mAnchorPoint->setCenter( mapAnchor );
+    mRubberScale = QPointF( rubberAnchor.x() - mRubberBand->x(), rubberAnchor.y() - mRubberBand->y() );
+    mRubberBand->setTransformOriginPoint( rubberAnchor );
+    mRubberBand->show();
+    mBaseDistance = toLayerCoordinates( mLayer, e->mapPoint() ).distance( mFeatureCenter );
+    mScaling = 1.0;
+
+    createScalingWidget();
+    if ( e->modifiers() & Qt::ShiftModifier )
+    {
+      if ( mScalingWidget )
+      {
+        mScalingWidget->setMagnet( 45 );
+      }
+    }
+
+    mScalingActive = true;
+
+    return;
+  }
+
+  applyScaling( mScaling );
+}
+
+void QgsMapToolScaleFeature::cancel()
+{
+  deleteScalingWidget();
+  deleteRubberband();
+  QgsVectorLayer *mLayer = currentVectorLayer();
+  if ( mLayer->selectedFeatureCount() == 0 )
+  {
+    mAnchorPoint.reset();
+  }
+  mScalingActive = false;
+}
+
+void QgsMapToolScaleFeature::updateRubberband( double scale )
+{
+  if ( mScalingActive )
+  {
+    mScaling = scale;
+
+    double offsetx = ( 1 - mScaling ) * mRubberScale.x();
+    double offsety = ( 1 - mScaling ) *  mRubberScale.y();
+
+    if ( mRubberBand )
+    {
+      mRubberBand->setTransform( QTransform( mScaling, 0, 0, mScaling, offsetx, offsety ) );
+      mRubberBand->update();
+    }
+  }
+}
+
+
+void QgsMapToolScaleFeature::applyScaling( double scale )
+{
+  mScaling = scale;
+  mScalingActive = false;
+
+  QgsVectorLayer *mLayer = currentVectorLayer();
+  if ( !mLayer )
+  {
+    deleteRubberband();
+    notifyNotVectorLayer();
+    return;
+  }
+
+  //calculations for affine transformation
+
+  mLayer->beginEditCommand( tr( "Features Scaled" ) );
+
+  int start = ( mLayer->geometryType() == 2 )? 1 : 0;
+
+  int i = 0;
+  const auto constMScaledFeatures = mScaledFeatures;
+  for ( QgsFeatureId id : constMScaledFeatures )
+  {
+    QgsFeature feat;
+    mLayer->getFeatures( QgsFeatureRequest().setFilterFid( id ) ).nextFeature( feat );
+    QgsGeometry geom = feat.geometry();
+    i = start;
+    QgsPointXY vertex = geom.vertexAt( i );
+    while ( !vertex.isEmpty() )
```

Comment:
Much more efficient would be to construct a QTransform and use QgsGeometry::transform here, instead of the vertex by vertex manipulation.

Reply:
Qtransform seems risky, but using the currend transform could be doable.

Reply:
Tried it, get succes but geometry didn't change...unsure why it doesn't work. Will upload other fixes once I figure out why it fails or revert it to the vertex approach.

Reply:
Reverted to vertex, commented out transform approach until it works.